File: _1_Hashmap/Session0.java
package _1_Hashmap;

import java.util.HashMap;
import java.util.Map;

public class Session0 {
    public static void main(String[] args) {
        int[] arr = {1, 2, 3, 4, 5, 6, 7, 8, 9, 10};

        // // Brute Force
        // int max = arr[0];
        // int min = arr[0];

        // for (int i = 1; i < arr.length; i++) {
        //     if (arr[i] > max) {
        //         max = arr[i];
        //     }
        //     if (arr[i] < min) {
        //         min = arr[i];
        //     }
        // }

        // System.out.println("Max: " + max);
        // System.out.println("Min: " + min);

        // Optimal Approach
        HashMap<Integer, Integer> map = new HashMap<>();
        for(int num: arr){
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        int minFreq = Integer.MAX_VALUE;
        int maxFreq = Integer.MIN_VALUE;
        int min = -1;
        int max = -1;
        // for(int key: map.keySet()){
        //     int freq = map.get(key);
        //     if(freq < minFreq){
        //         minFreq = freq;
        //         min = key;
        //     }
        //     if(freq > maxFreq){
        //         maxFreq = freq;
        //         max = key;
        //     }
        // }
        // System.out.println("Max: " + max);
        // System.out.println("Min: " + min);

        for(Map.Entry<Integer, Integer> entry: map.entrySet()){
            int key = entry.getKey();
            int freq = entry.getValue();
            if(freq < minFreq){
                minFreq = freq;
                min = key;
            }
            if(freq > maxFreq){
                maxFreq = freq;
                max = key;
            }
        }
        System.out.println("Max: " + max);
        System.out.println("Min: " + min);


    }
}

File: _1_Hashmap/Session0.py
# HashMap equivalent in Python is a dictionary

def main():
    arr = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

    # # Brute Force
    # max_val = arr[0]
    # min_val = arr[0]

    # for i in range(1, len(arr)):
    #     if arr[i] > max_val:
    #         max_val = arr[i]
    #     if arr[i] < min_val:
    #         min_val = arr[i]

    # print("Max:", max_val)
    # print("Min:", min_val)

    # Optimal Approach using dictionary
    freq_map = {}

    for num in arr:
        freq_map[num] = freq_map.get(num, 0) + 1

    min_freq = float('inf')
    max_freq = float('-inf')
    min_key = -1
    max_key = -1

    for key, freq in freq_map.items():
        if freq < min_freq:
            min_freq = freq
            min_key = key
        if freq > max_freq:
            max_freq = freq
            max_key = key

    print("Max:", max_key)
    print("Min:", min_key)


if __name__ == "__main__":
    main()
